lib/Conversion/PipelineToHW/PipelineToHW.cpp

```diff
@@ -362,89 +463,93 @@ class PipelineOutlineLowering : public PipelineLowering {
     // operations are inserted into the pipeline module.
     builder.setInsertionPointToStart(pipelineMod.getBodyBlock());
 
-    PipelineStageMod pipelineStageMod(*this, pipeline.getEntryStage(),
-                                      pipelineMod, withStall);
-
-    if (failed(lowerStage(pipeline.getEntryStage(), pipelineStageMod.inputs,
-                          pipelineStageMod.valid, pipelineStageMod.stall, 0)))
+    pipelineStageMod =
+        PipelineStageMod(*this, pipeline.getEntryStage(), pipelineMod,
+                         withStall, /*isParentPipeline*/ true);
+
+    StageArgs args;
+    args.data = pipelineStageMod.inputs;
+    args.enable = pipelineStageMod.enable;
+    args.clock = pipelineStageMod.clock;
+    args.reset = pipelineStageMod.reset;
+    args.stall = pipelineStageMod.stall;
+    FailureOr<StageReturns> lowerRes =
+        lowerStage(pipeline.getEntryStage(), args, 0);
+    if (failed(lowerRes))
       return failure();
 
+    // Assign the output op of the top-level pipeline module.
+    auto outputOp =
+        cast<hw::OutputOp>(pipelineMod.getBodyBlock()->getTerminator());
+    outputOp->setOperands(lowerRes->passthroughs);
+    outputOp->insertOperands(lowerRes->passthroughs.size(), lowerRes->valid);
     pipeline.erase();
+
+    // Mini-optimization: the lowerStage/emitBody logic is written to not do any
+    // special-case logic for the final stage of the pipeline.
+    // In many cases, there are no operations in the final stage, except for the
+    // return op. In these cases, we'll just inline the exit stage into the
+    // pipeline module, and erase the (empty) last stage module.
+    inlineIfEmpty(currentStageInst);
+
     return success();
   }
 
   /// NOLINTNEXTLINE(misc-no-recursion)
```

Comment:
Could you re-write this to not be recursive? Seems easy enough given that it's tail-recursive. (Though I feel like I've made this comment before and that there's some subtly which makes it non-trivial.)